### DISCUSSION TITLE
Install Aria2 on aio nextcloud container

### DISCUSSION THREAD

**Discussion by securegh** [nextcloud/all-in-one#1159] [Ideas] [answered]
### How to use GitHub
* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Is your feature request related to a problem? Please describe. 

I use **NCDownloader**,  the app relies on **Aria2**,  I have to manually add aria2 package to the nextcloud container to make NCDowloader app work. I have to manually add Aria2 package whenever the container is rebooted or updated.

### Describe the solution you'd like

It would be awesome to have **Aria2** package installed by default on the aio nextcloud container.

**Comment by szaimen** [comment]
I an sorry but we cannot add external dependencies for each and every app into the container as it would become unmaintainable very fast.

**Reply by Zoey2936** [reply]
Maybe, there could be an option, that you could set over an env for example that on every container start custom apk packages are being installed in the nextcloud container? Just as an Idea.

**Reply by securegh** [reply]
Thanks for the reply! 

Is there a persistant way to install the aria2 package during container startup ? 
Perhaps a persistant bootup script that I can add that install cmd line?  

Thank you!

**Reply by szaimen** [reply]
> Maybe, there could be an option, that you could set over an env for example that on every container start custom apk packages are being installed in the nextcloud container? Just as an Idea.

I fear that this will not work as the nextcloud container is not running as root user and cannot install packages. So the process would need to come externally via docker exec from the mastercontainer which is a really hacky solution and not desirable from a maintenance perspective...

**Reply by securegh** [reply]
Is there a way for me to modify the mastercontainer bootup script for nextcloud container to add that specific package? 

Thanks!

**Reply by szaimen** [reply]
I'll think about a solution for adding external dependencies as I find the idea interesting...

**Reply by szaimen** [reply]
Will get handled in https://github.com/nextcloud/all-in-one/issues/1162

**Reply by securegh** [reply]
Thank you for considering this issue, I hope you find the best way for adding external dependencies.

**Comment by szaimen** [comment, marked as answer]
This will be handled with https://github.com/nextcloud/all-in-one/issues/1162

**Reply by szaimen** [reply]
This is now released with v3.0.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Reply by securegh** [reply]
That's awesome! Thank you for implementing this feature so fast! 
I try it out and get back to you!

**Reply by securegh** [reply]
I am using the beta version, but I don't know how this update implement adding dependencies. 
Does this version adds by default aria2 to the nextcloud container or I am support to add the dependencies I want into a config file? 
Right now, aria2 **is not added by default** to nextcloud container. 
Please let me know how the beta version implement adding dependencies, thanks!

**Reply by szaimen** [reply]
See https://github.com/nextcloud/all-in-one#how-to-add-packets-permanently-to-the-nextcloud-container

**Reply by securegh** [reply]
Awesome! It works very well, aria2 is installed automatically after nextcloud container startup. 
Thank you for making this feature a reality this fast, it speaks volume on the team effectiveness and effort to make this project the best. 
I really like NC AiO project, I am available for any beta testing your team needs.  
Thanks again!

**Reply by martin-77** [reply]
Great to hear, that this gets implemented.
Switched to beta and tried to start aria2 but ran into this error:

```
The command "'/var/www/html/custom_apps/ncdownloader/lib/Aria2/../../bin/aria2c' '--continue' '--daemon=true' '--enable-rpc=true' '--rpc-secret=ncdownloader123' '--listen-port=51413' '--rpc-listen-port=6800' '--follow-torrent=true' '--enable-dht=true' '--enable-peer-exchange=true' '--peer-id-prefix=-TR2770-' '--user-agent=Transmission/2.77' '--log-level=notice' '--seed-ratio=1.0' '--bt-seed-unverified=true' '--max-connection-per-server=4' '--max-concurrent-downloads=10' '--check-certificate=false' '--on-download-complete=/var/www/html//apps/ncdownloader/hooks/completeHook.sh' '--on-download-start=/var/www/html//apps/ncdownloader/hooks/startHook.sh' '--save-session=/mnt/ncdata/aria2/aria2.session' '--input-file=/mnt/ncdata/aria2/aria2.session' '--log=/mnt/ncdata/aria2/aria2.log'" failed.

Exit Code: 2(Misuse of shell builtins)

Working directory: /var/www/html

Output:
================


Error Output:
================
/var/www/html/custom_apps/ncdownloader/lib/Aria2/../../bin/aria2c: line 1: syntax error: unexpected "("

```
@securegh could you explain, how you made it work? Or could you post your docker-compose.yml ? 

Thanks in advance!

**Reply by Zoey2936** [reply]
Just Go to the nextcloud downloader settings and adjust the paths
![grafik](https://user-images.githubusercontent.com/75573284/202496517-9c1ab36c-e894-491d-a903-82249e118b7f.png)

**Reply by securegh** [reply]
@martin-77 
I got similar error before! First make sure you have aria2 install on the nextcloud container 

`docker exec -it nextcloud-aio-nextcloud apk info | grep aria2`

Then I usually I disable and remove NCDowloader app and reinstall it and that would fix the error for me. 

Here is my docker run script for NC AIO, it's running behind a reverse proxy. 

```
#! /bin/bash

docker run -it \
--sig-proxy=false \
--name nextcloud-aio-mastercontainer \
--restart always \
--publish 8080:8080 \
-e APACHE_PORT=11000 \
-e NEXTCLOUD_ADDITIONAL_APKS="aria2 imagemagick" \
--volume nextcloud_aio_mastercontainer:/mnt/docker-aio-config \
--volume /var/run/docker.sock:/var/run/docker.sock:ro \
nextcloud/all-in-one:beta-arm64

```

Hope that helps!
